pysmt/oracles.py

```diff
@@ -319,6 +319,37 @@ def walk_div(self, formula, args, **kwargs):
         theory_out = theory_out.set_difference_logic(False)
         return theory_out
 
+    def walk_floordiv(self, formula, args, **kwargs):
```

Comment:
Technically, floordiv is non part of LIA even when the divisor is an integer constant. Consider for example `x // 5 == y` where x and y are integers, this is *not* equivalent to `x == y * 5` because `x=11` and `y=2` is a model of the first equation but not of the other. Moreover both floordiv and modulus are not part of SMTLib, but Z3-specific extensions. 

I think it would make sense to either mark the use of these operators as a NIA formula (and Z3 is the only solver that can currently deal with it) or create a special logic for LIA + floordiv + mod. 

Alternatively, for the 'quasi-linear' case, one could define the floordiv in terms of modulus:
```
(define-fun floordiv ((x Int) (y Int)) Int
  (ite (or (= (mod x y) 0) (>= x 0))
       (div x y)
       (ite (>= y 0)
            (+ (div x y) 1)
            (- (div x y) 1))))
```
This definition should be supported also by MathSAT via the modular congruence.

@marcogario @agriggio Any thoughts? What would be a clean way of including this logic for solvers that supports it?

Reply:
For what concerns logic extensions outside the SMT-LIB, we already have some precedent in the theory of array with constant element. In that case we created a custom bit in the Theory description, and thus defined logics with this extended concept. I would follow the same approach here.